{% extends 'wrapper.html' %}
{% block body %}

    <link href='static/style/about.css' rel='stylesheet' type='text/css'/>
    <script src="static/js/lib/highcharts.js"></script>
    <script src="static/js/stats.js"></script>

    <h1>About</h1>
    <section>
        <p>Gregory W. Gundersen, Matthew R. Jones, Andrew D. Rouillard, Yan Kou, Caroline D. Monteiro, Axel S. Fledmann, Kevin S. Hu, Avi Ma’ayan. GEO2Enrichr: browser extension and server app to extract gene sets from GEO and analyze them for biological functions. <a href="http://www.ncbi.nlm.nih.gov/pubmed/25971742" target="_blank">Bioinformatics. 2015 Sep 15;31(18):3060-2. doi: 10.1093</a>.</p>
    </section>
    <section>
        <h2>Summary</h2>
        <p>GEO2Enrichr is a web application and two browser extensions (one for Chrome and another for Firefox) designed to ease the extraction of signatures from studies posted on the <a href='http://www.ncbi.nlm.nih.gov/geo/' target='_blank'>Gene Expression Omnibus (GEO)</a> database. These signatures are then submitted to <a href='http://amp.pharm.mssm.edu/Enrichr/' target='_blank'>Enrichr</a>, a gene signature search engine containing over 70,000 annotated signatures, for downstream functional analysis. GEO2Enrichr is <a href='https://github.com/MaayanLab/geo2enrichr' target='_blank'>open source</a> and being actively developed by the <a href='http://icahn.mssm.edu/research/labs/maayan-laboratory' target='_blank'>Ma'ayan Laboratory</a> at the <a href='http://icahn.mssm.edu/' target='_blank'>Icahn School of Medicine at Mount Sinai</a> as a part of the efforts of the <a href='http://www.lincs-dcic.org/' target='_blank'>BD2K-LINCS-DCIC</a> and <a href='http://commonfund.nih.gov/idg/index' target='_blank'>KMC-IDG</a>, which are <a href="https://commonfund.nih.gov/ " target="_blank">NIH Common Fund</a> sponsored centers.</p>
    </section>
    <section>
        <h2>Abstract</h2>
        <p>
            <strong>Motivation:</strong> Identification of differentially expressed genes is an important step in extracting knowledge from gene expression profiling studies. The raw expression data from microarray and other high-throughput technologies is deposited into the Gene Expression Omnibus (GEO) and served as Simple Omnibus Format in Text (SOFT) files. However, to extract and analyze differentially expressed genes from GEO requires significant computational skills. Here we introduce GEO2Enrichr, a browser extension for extracting differentially expressed gene sets from GEO and analyzing those sets with Enrichr, an independent gene set enrichment analysis tool containing over 70,000 annotated gene sets organized into 75 gene-set libraries. GEO2Enrichr adds JavaScript code to GEO web-pages; this code scrapes user selected accession numbers and metadata, and then, with one click, users can submit this information to a web-server application that downloads the SOFT files, parses, cleans and normalizes the data, identifies the differentially expressed genes, and then pipes the resulting lists to Enrichr for downstream functional analysis. GEO2Enrichr opens a new avenue for adding functionality to major bioinformatics resources such GEO by integrating tools and resources without the need for a plug-in architecture. Importantly, GEO2Enrichr helps researchers to quickly explore hypotheses with little technical overhead, lowering the barrier of entry for biologists by automating data processing steps needed for knowledge extraction from the major repository GEO.
        </p>
        <p>
        <strong>Availability:</strong> GEO2Enrichr is an <a href='https://github.com/MaayanLab/geo2enrichr' target='_blank'>open source</a> tool, freely available for installation as browser extensions at the Chrome Web Store and FireFox Add-ons. Documentation and a browser independent web application can be found on this website.
        </p>
    </section>

    <h1>Stats</h1>
    <section>
        <div>
            <table>
                <tr>
                    <td>Gene signatures</td>
                    <td>{{ stats.num_gene_signatures }}</td>
                </tr>
                <tr>
                    <td>Gene lists</td>
                    <td>{{ stats.num_gene_lists }}</td>
                </tr>
                <tr>
                    <td>Tags</td>
                    <td>{{ stats.num_tags }}</td>
                </tr>
                <tr>
                    <td>Platforms</td>
                    <td>{{ stats.num_platforms }}</td>
                </tr>
            </table>

        </div>
        <div id="platforms-bar-chart"></div>
    </section>

    <script>
        $(function() {
            buildStats({{ stats_json|safe }});
        });
    </script>

{% endblock %}