<section id='summary'>
    <h2 class='first'>Summary</h2>
    <p>GEO2Enrichr is a <a href='http://en.wikipedia.org/wiki/Browser_extension' target='_blank'>Google Chrome Extension</a> and Python server developed to ease the extraction of signatures from studies posted on the <a href='http://www.ncbi.nlm.nih.gov/geo/' target='_blank'>Gene Expression Omnibus (GEO)</a> database. These signatures are then submitted to <a href='http://amp.pharm.mssm.edu/Enrichr/' target='_blank'>Enrichr</a>, a gene signature search engine containing over 70,000 annotated signatures, for downstream functional analysis. GEO2Enrichr is open source and will be hosted on <a href='https://github.com/' target='_blank'>GitHub</a> upon publication acceptance of an article describing the software. GEO2Enrichr is being actively developed by the <a href='http://icahn.mssm.edu/research/labs/maayan-laboratory' target='_blank'>Ma'ayan Laboratory</a> at the <a href='http://icahn.mssm.edu/' target='_blank'>Icahn School of Medicine at Mount Sinai</a> as a part of the efforts of the <a href='http://www.lincs-dcic.org/' target='_blank'>BD2K-LINCS-DCIC</a> and <a href='http://commonfund.nih.gov/idg/fundedresearch' target='_blank'>KMC-IDG NIH Common Fund</a> sponsored centers.</p>
</section>
<section id='abstract'>
    <h2>Abstract</h2>
    <p>Identification of differentially expressed genes from gene expression profiling studies is a necessary and important step in the analysis of microarray and RNA-seq data. Such data is deposited into the Gene Expression Omnibus (GEO). However, currently, to extract and analyze differentially expressed genes from GEO requires significant computational expertise. GEO2Enrichr has a front-end browser extension and back-end server. The back-end is written in Python and uses the Flask framework for the web application server. The server exposes three API endpoints to: (1) query GEO; (2) identify the differentially expressed genes; and (3) query Enrichr. The API endpoints are chained into a useful data pipeline. The front-end is a Google Chrome Extension written in JavaScript and CSS and programmatically inserts HTML elements into the native GEO interface. The Chrome web browser loads the extension whenever the user navigates to an NCBI GEO webpage. GEO2Enrichr unobtrusively inserts itself onto the page as a single button and checkboxes for easy sample selection. The button opens a modal box that allows users to edit their selected data, choose settings for differential expression identification, insert metadata about the experiments, download their resultant gene lists, and submit their gene lists to Enrichr for further analysis. Differential expression analysis has five options: Characteristic Direction, T-test, limma, SAM and fold-change. Other options are available to control for the size of the resultant differentially expressed gene lists by setting p-value and FDR cutoffs. GEO2Enrichr is a Google Chrome Extension and Python-based API that adds functionality to GEO by allowing users to pipe Simple Omnibus Format in Text (SOFT) files to a differential expression analysis tool, and then pipe the differentially expressed genes for analysis with Enrichr, a popular gene list enrichment analysis web application. GEO2Enrichr addresses a need for improved functionality of GEO by embedding new features into the existing GEO pages, allowing researchers to easily select samples and process them for differential expression by various methods and perform enrichment analyses. GEO2Enrichr is free and available for installation at the Chrome Web Store. GEO2Enrichr can facilitate the more broad reusability of the GEO resource by lowering the point of entry to biologists without computational expertise. The systematic use of GEO2Enrichr can generate a new useful searchable resource. Overall, GEO2Enrichr can lead to improved extraction of knowledge from data.</p>
</section>
