<section>
    <h2>Data processing pipeline</h2>
    <p>GEO2Enrichr analyzes your data by performing the following operations:</p>
    <ol id="g2e-process">
        <li>Downloads the associated SOFT file from GEO or uploads a custom SOFT file via this website.</li>
        <li>Discards data with missing values or one-to-many probe-to-gene mappings.</li>
        <li>log2 transforms the data if necessary.</li>
        <li><a href='http://bioinformatics.oxfordjournals.org/content/19/2/185.full.pdf+html' target='_blank'>Quantile normalizes</a> the data if necessary.</li>
        <li>Averages multiple probe IDs to single gene symbols using a curated platform-to-probe-to-gene table.</li>
        <li>Identifies differentially expressed genes with the selected method. The default method is the <a href='http://maayanlab.net/CD/' target='_blank'>characteristic direction</a> method, developed by the Ma'ayan Laboratory and shown to outperform all other popular methods.</li>
        <li>Writes the resultant gene lists (up, down, combined) into plain text files for download.</li>
        <li>Pipes the gene lists to <a href='http://amp.pharm.mssm.edu/Enrichr/' target='_blank'>Enrichr</a> for further analysis.</li>
    </ol>
</section>
