<!DOCTYPE html>
<html>
<head>
    <title>GEO2Enrichr</title>
	<meta charset='UTF-8'>
	i
	<meta content='width=device-width, initial-scale=1' name='viewport'/>
    <link rel="shortcut icon" href="favicon.ico" type="image/x-icon">
    <link rel="icon" href="favicon.ico" type="image/x-icon">
	<link rel='stylesheet' type='text/css' href='css/main.css'/>
	<link rel='stylesheet' type='text/css' href='css/open-sans.css'/>
</head>
<body>
    <div id='sidebar'>
        <ul>
            <li><a href='#summary'>Summary</a></li>
            <li><a href='#abstract'>Abstract</a></li>
            <li><a href='#installation'>Installation</a></li>
            <li><a href='#tutorial'>Tutorial</a></li>
            <li><a href='#processing'>Data processing</a></li>
            <li><a href='#api'>API</a></li>
            <li><a href='#contact'>Contact</a></li>
        </ul>
    </div>
    <div id='content'>
        <h2 id='title'>
            <img id='logo' src='images/g2e-logo-50x50.png'/>
            <span>GEO2</span><span class='highlight'>Enrichr</span>
        </h2>
        <h4 id='subtitle'>Google Chrome Extension to extract gene sets from GEO and analyze these lists for common biological functions. Click <a href='https://chrome.google.com/webstore/detail/geo2enrichr/pcbdeobileclecleblcnadplfcicfjlp' target='_blank'>here</a> to install it.</h4>
        <section id='summary'>
            <h3 class='first'>Summary</h3>
            <p>GEO2Enrichr is a <a href='http://en.wikipedia.org/wiki/Browser_extension' target='_blank'>Google Chrome Extension</a> and Python server developed to ease the extraction of signatures from studies posted on the <a href='http://www.ncbi.nlm.nih.gov/geo/' target='_blank'>Gene Expression Omnibus (GEO)</a> database. These signatures are then submitted to <a href='http://amp.pharm.mssm.edu/Enrichr/' target='_blank'>Enrichr</a>, a gene signature search engine containing over 70,000 annotated signatures, for downstream functional analysis. GEO2Enrichr is open source and will be hosted on <a href='https://github.com/' target='_blank'>GitHub</a> upon publication acceptance of an article describing the software. GEO2Enrichr is being actively developed by the <a href='http://icahn.mssm.edu/research/labs/maayan-laboratory' target='_blank'>Ma'ayan Laboratory</a> at the <a href='http://icahn.mssm.edu/' target='_blank'>Icahn School of Medicine at Mount Sinai</a> as a part of the efforts of the <a href='http://www.lincs-dcic.org/' target='_blank'>BD2K-LINCS-DCIC</a> and <a href='http://commonfund.nih.gov/idg/fundedresearch' target='_blank'>KMC-IDG NIH Common Fund</a> sponsored centers.</p>
        </section>
        <section id='abstract'>
            <h3>Abstract</h3>
            <p>Identification of differentially expressed genes from gene expression profiling studies is a necessary and important step in the analysis of microarray and RNA-seq data. Such data is deposited into the Gene Expression Omnibus (GEO). However, currently, to extract and analyze differentially expressed genes from GEO requires significant computational expertise. GEO2Enrichr has a front-end browser extension and back-end server. The back-end is written in Python and uses the Flask framework for the web application server. The server exposes three API endpoints to: (1) query GEO; (2) identify the differentially expressed genes; and (3) query Enrichr. The API endpoints are chained into a useful data pipeline. The front-end is a Google Chrome Extension written in JavaScript and CSS and programmatically inserts HTML elements into the native GEO interface. The Chrome web browser loads the extension whenever the user navigates to an NCBI GEO webpage. GEO2Enrichr unobtrusively inserts itself onto the page as a single button and checkboxes for easy sample selection. The button opens a modal box that allows users to edit their selected data, choose settings for differential expression identification, insert metadata about the experiments, download their resultant gene lists, and submit their gene lists to Enrichr for further analysis. Differential expression analysis has five options: Characteristic Direction, T-test, limma, SAM and fold-change. Other options are available to control for the size of the resultant differentially expressed gene lists by setting p-value and FDR cutoffs. GEO2Enrichr is a Google Chrome Extension and Python-based API that adds functionality to GEO by allowing users to pipe Simple Omnibus Format in Text (SOFT) files to a differential expression analysis tool, and then pipe the differentially expressed genes for analysis with Enrichr, a popular gene list enrichment analysis web application. GEO2Enrichr addresses a need for improved functionality of GEO by embedding new features into the existing GEO pages, allowing researchers to easily select samples and process them for differential expression by various methods and perform enrichment analyses. GEO2Enrichr is free and available for installation at the Chrome Web Store. GEO2Enrichr can facilitate the more broad reusability of the GEO resource by lowering the point of entry to biologists without computational expertise. The systematic use of GEO2Enrichr can generate a new useful searchable resource. Overall, GEO2Enrichr can lead to improved extraction of knowledge from data.</p>
        </section>
        <section id='installation'>
            <h3>Installation</h3>
            <p>Click <a href='https://chrome.google.com/webstore/detail/geo2enrichr/pcbdeobileclecleblcnadplfcicfjlp' target='_blank'>here</a> to install GEO2Enrichr as an extension in your Chrome browser.</p>
            <p>Clicking the <strong>+FREE</strong> button will install and activate the application. The Chrome Web Store will ask you to accept GEO2Enrichr's permission settings. The GEO2Enrichr application needs permission to make HTTP requests to our server for processing the dataset you will select when interacting with entries listed on the GEO site.</p>
        </section>
        <section id='tutorial'>
            <h3>Tutorial</h3>
            <p>GEO2Enrichr works by embedding additional functionality into relevant GEO pages that you visit. For example, after installing the Chrome Extension, visiting the page <a href='http://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS5077' target='_blank'>GDS5077</a> and clicking "Compare 2 sets of samples", you should see a new hypertext that says <strong>"Pipe to Enrichr"</strong> (Fig. 1).</p>
            <img src='images/screenshot-embedded.png'/>
            <p class='caption'>Fig. 1 Screenshot that shows how GEO2Enrichr adds an additional step to the GEO datasets pipeline.</p>
            <p>If you immediately click that hypertext, you will receive a warning: "Please select 2 or more control samples." This warning occurs because GEO2Enrichr needs you to specify the datasets that you would like to download, process, and analyze. Click <strong>"Select which Samples to put in Group A and Group B"</strong> and select two groups of samples: control samples on the left and treatment/condition samples on the right. In this example, GSM1071454 and GSM1071455 are the control samples and GSM1071457 and GSM1071457 are the treatment or condition samples (Fig. 2). It is important that the control samples are selected on the left and treatment samples are selected on the right because this information is used to determine the up-regulated and down-regulated genes direction.</p>
            <img src='images/screenshot-samples.png'/>
            <p class='caption'>Fig. 2 Screenshot to show how to identify the control and treatment/condition samples on the GEO datasets site.</p>
            <p>Now you can open GEO2Enrichr. Click <strong>Pipe to Enrichr</strong> to see the application's main modal dialog box (Fig. 3).</p>
            <img src='images/screenshot-modal.png'/>
            <p class='caption'>Fig. 3 Screenshot of the GEO2Enrichr main modal dialog box.</p>
            <p>The GEO2Enrichr application <a href='http://en.wikipedia.org/wiki/Data_scraping#Screen_scraping' target='_blank'>screen scrapes</a> the GEO webpage for relevant data including the data the user selected. Please confirm that the accession number, platform, organism, and sample files are all correct. You can edit the first three fields by clicking on their text. If you need to change the sample selection, you must close GEO2Enrichr and re-select the samples via the GEO dataset interactive interface (Step 2).</p>
            <p>Please fill in any metadata about the experiment, such as cell type or tissue, perturbation, and manipulated gene (if relevant). This metadata will be used to annotate your downloadable filenames and a description that uses this information will be sent to Enrichr.</p>
            <p>Finally, click the <strong>Submit samples</strong> button to submit your data selection to GEO2Enrichr for data processing. Depending on the size of SOFT files and the number of samples you selected, this can take anywhere from a few seconds to a minute or two. A progress bar will keep you updated on the progress. Once the processing is complete, the output will show three links to Enrichr, one for up genes, one for down genes, and one for both up and down combined. In addition to these buttons, you will be provided with a button to download your gene lists as text files.</p>
        </section>
        <section id='processing'>
            <h3>Data processing</h3>
            <p>Given sample selection accession numbers, GEO2Enrichr analyzes these samples by performing the following operations:</p>
            <ol id="g2e-process">
                <li>Downloads the associated SOFT file from GEO.</li>
                <li>Discards data with missing values or one-to-many probe-to-gene mappings.</li>
                <li>log2 transforms the data if necessary.</li>
                <li><a href='http://en.wikipedia.org/wiki/Quantile_normalization' target='_blank'>Quantile normalizes</a> the data if necessary.</li>
                <li>Averages multiple probes to single genes.</li>
                <li>Identifies differentially expressed genes with the selected method. The default method is the <a href='http://maayanlab.net/CD/' target='_blank'>characteristic direction</a> method, developed by the Ma'ayan Laboratory and shown to outperform all other popular methods.</li>
                <li>Writes the top and bottom differentially expressed genes into text.</li>
                <li>Pipes the gene lists to <a href='http://amp.pharm.mssm.edu/Enrichr/' target='_blank'>Enrichr</a> for further analysis.</li>
	        </ol>
        </section>
        <section id='api'>
            <h3>API</h3>
            <p>GEO2Enrichr can be accessed programmatically using APIs.</p>
            <p>All requests should go to the <span class='underline'>http://amp.pharm.mssm.edu/g2e/full</span> endpoint:</p>
            <!-- http://localhost:8083/g2e/full?accession=GDS5077&platform=GPL10558&control=GSM1071454-GSM1071455&experimental=GSM1071457-GSM1071456 -->
            <table>
                <caption>Required arguments</caption>
                <tr>
                    <td>accession</td>
                    <td>A valid <a href='http://en.wikipedia.org/wiki/Accession_number_%28bioinformatics%29' target='_blank'>accession number</a> for GEO, e.g. <span class='mono'>accession=GDS5077</span>.</td>
                </tr>
                <tr>
                    <td>platform</td>
                    <td>The associated platform for a given accession number, e.g. <span class='mono'>platform=GPL10558</span>.</td>
                </tr>
                <tr>
                    <td>control</td>
                    <td>A hyphen-separated list of GSM files, e.g.<span class='mono'>control=GSM1071454-GSM1071455</span></td>
                </tr>
                <tr>
                    <td>experimental</td>
                    <td>A hyphen-separated list of GSM files, e.g.<span class='mono'>experimental=GSM1071457-GSM1071456</span></td>
                </tr>
            </table>
            <table>
                <caption>Optional arguments</caption>
                <tr>
                    <td>organism</td>
                    <td>An organism, e.g. <span class='mono'>organism=homo+sapiens</span>.<em> Note that GEO2Enrichr and Enrichr currently only support human and mouse genes</em>.</td>
                </tr>
                <tr>
                    <td>cell</td>
                    <td>The cell or tissue type used in the experiment, e.g. <span class='mono'>cell=RUES2</span>.</td>
                </tr>
                <tr>
                    <td>perturbation</td>
                    <td>A perturbation, e.g. <span class='mono'>perturbation=depleted+TMEM88</span></td>
                </tr>
                <tr>
                    <td>gene</td>
                    <td>If relevant, the manipulated gene.</td>
                </tr>
            </table>

            <table>
                <caption>Example</caption>
                <tr>
                    <td>Valid request</td>
                    <td><span class='mono'>http://amp.pharm.mssm.edu/g2e/full?accession=GDS5077&platform=GPL10558&control=GSM1071454-GSM1071455&experimental=GSM1071457-GSM1071456</span></td>
                </tr>
            </table>
        </section>
        <section id='contact'>
            <h3>Contact</h3>
            <p>Please contact Avi Ma’ayan and Gregory Gundersen for comments, suggestions, and support: <a href='mailto:avi.maayan@mssm.edu'>avi.maayan@mssm.edu</a>, <a href='mailto:gregory.gundersen@mssm.edu'>gregory.gundersen@mssm.edu</a>.</p>
        </section>
    </div>
</body>
</html>
